File: libs/common/common/ids.py
from __future__ import annotations

from typing import NewType

from common.utils import TSID

RequestId = NewType("RequestId", TSID)
SqsMessageId = NewType("SqsMessageId", TSID)
UserId = NewType("UserId", TSID)
GameId = NewType("GameId", TSID)
GameEventId = NewType("GameEventId", TSID)
PlayerId = NewType("PlayerId", TSID)
AgentId = NewType("AgentId", TSID)
AgentVersionId = NewType("AgentVersionId", TSID)
AgentVersionToolId = NewType("AgentVersionToolId", TSID)
AgentStatisticsId = NewType("AgentStatisticsId", TSID)
AgentExecutionSessionId = NewType("AgentExecutionSessionId", TSID)
AgentIterationHistoryId = NewType("AgentIterationHistoryId", TSID)
TestScenarioId = NewType("TestScenarioId", TSID)
TestScenarioResultId = NewType("TestScenarioResultId", TSID)
ToolId = NewType("ToolId", TSID)
LLMIntegrationId = NewType("LLMIntegrationId", TSID)
LLMUsageId = NewType("LLMUsageId", TSID)
EventId = NewType("EventId", TSID)
ErrorReportId = NewType("ErrorReportId", TSID)

File: client/src/types/ids.ts
/**
 * Tagged string types for all IDs to prevent mixing different ID types
 * and ensure type safety throughout the frontend.
 */

// Base branded type utility
type Brand<T, K> = T & { __brand: K };

// ID Types matching backend TSID format
export type AgentId = Brand<string, 'AgentId'>;
export type AgentVersionId = Brand<string, 'AgentVersionId'>;
export type PlayerId = Brand<string, 'PlayerId'>;
export type ToolId = Brand<string, 'ToolId'>;
export type UserId = Brand<string, 'UserId'>;
export type GameId = Brand<string, 'GameId'>;
export type TestScenarioId = Brand<string, 'TestScenarioId'>;
export type AgentStatisticsId = Brand<string, 'AgentStatisticsId'>;
export type AgentExecutionSessionId = Brand<string, 'AgentExecutionSessionId'>;
export type AgentIterationHistoryId = Brand<string, 'AgentIterationHistoryId'>;
export type LLMIntegrationId = Brand<string, 'LLMIntegrationId'>;
export type EventId = Brand<string, 'EventId'>;
export type ErrorReportId = Brand<string, 'ErrorReportId'>;